session.go

```diff
@@ -0,0 +1,233 @@
+package apns
+
+import (
+	"container/list"
+	"errors"
+	"io"
+	"sync"
+)
+
+type SessionError struct {
+	Notification Notification
+	Err          Error
+}
+
+func (s SessionError) Error() string {
+	return s.Err.Error()
+}
+
+type Session interface {
+	Send(n Notification) error
+	Connect() error
+	RequeueableNotifications() []Notification
+	Disconnect()
+	Disconnected() bool
+}
+
+type buffer struct {
+	size int
+	m    sync.Mutex
+	*list.List
+}
+
+func newBuffer(size int) *buffer {
+	return &buffer{size, sync.Mutex{}, list.New()}
+}
+
+func (b *buffer) Add(v interface{}) *list.Element {
+	b.m.Lock()
+	defer b.m.Unlock()
+
+	e := b.PushBack(v)
+
+	if b.Len() > b.size {
+		b.Remove(b.Front())
+	}
+
+	return e
+}
+
+type sessionState int
+
+const (
+	sessionStateNew          sessionState = iota
+	sessionStateConnected    sessionState = iota
+	sessionStateDisconnected sessionState = iota
+)
+
+type session struct {
+	b *buffer
+
+	conn  Conn
+	connm sync.Mutex
+
+	st  sessionState
+	stm sync.Mutex
+
+	id  uint32
+	idm sync.Mutex
+
+	err SessionError
+}
+
+func NewSession(conn Conn) Session {
+	return &session{
+		st:    sessionStateNew,
+		stm:   sync.Mutex{},
+		conn:  conn,
+		connm: sync.Mutex{},
+		idm:   sync.Mutex{},
+		b:     newBuffer(50),
```

Comment:
> There is also a cleanup thread running that constantly checks the oldest notification in the ‘Sent’ queue to see if it’s older than a few seconds and if so, it is assumed to have been successfully sent and is removed from the ‘Sent’ queue.  - [The Problem With Apples Push Notification Service](http://redth.codes/the-problem-with-apples-push-notification-ser/)

> It can take a while for the dropped connection to make its way from APNs back to your server just because of normal latency. It's possible to send over **500** notifications before a write fails because of the connection being dropped. Around **1,700** notifications writes can fail just because the pipe is full, so just retry in that case once the stream is ready for writing again. - [Push Notification Throughput and Error Checking](https://developer.apple.com/library/ios/technotes/tn2265/_index.html#//apple_ref/doc/uid/DTS40010376-CH1-TNTAG44)